src/coord/src/timestamp.rs

```diff
@@ -1866,3 +1893,72 @@ impl Timestamper {
         self.current_timestamp = new_ts;
     }
 }
+
+fn rt_kafka_metadata_fetch_loop(c: RtKafkaConnector, consumer: BaseConsumer, wait: Duration) {
+    debug!(
+        "Starting realtime Kafka thread for {} (source {})",
+        &c.topic, &c.id
+    );
+
+    let mut high_watermarks = vec![]; // high watermarks for each partition, in order
+    let mut i: usize = 0;
+    while !c.state.stop.load(Ordering::SeqCst) {
+        // Fetch metadata to check for new partitions. We do this less
+        // frequently than every tick because partitions are added very
+        // infrequently in production. We do this a bit more frequently than is
+        // ideal in order to speed up tests that dynamically add partitions.
+        if high_watermarks.len() == 0 || i % 10 == 0 {
+            match get_kafka_partitions(&consumer, &c.topic, Duration::from_secs(15)) {
+                Ok(partitions) => {
+                    if partitions.len() != high_watermarks.len() {
```

Comment:
I don't know if this can ever happen, but we could add an assertion here confirming that we "never" go backwards and return fewer partitions (for instance if the topic is deleted and recreated).

Reply:
Great idea, thanks. Done. I added a log message in that case. Ideally we might want to mark the source as permanently errored, but I'll punt that to future work.